LanguageExt.Core/DataTypes/Recursive/Recursive.cs

```diff
@@ -0,0 +1,68 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace LanguageExt.DataTypes.Recursive
+{
+    /// <summary>
+    /// While the .NET CLR supports tail call optimisation, the C# designers made the decision to limit the depth of the call stack.
+    /// This class allows recursive functions to have a near unlimited stack depth.    
+    /// This class is implemented using trampoling and implicit type conversion.   
+    /// </summary>    
+    /// <remarks>
+    /// Intended usage is for situations where an algorithm might more conveniently be expressed as 
+    /// a recursive function but fears this may result in a stack overflow error.
+    ///     
+    /// Note: This is a lazy type. The function is only invoked when an implicit cast occurs    
+    /// </remarks>
+    /// <typeparam name="A">Base case return type</typeparam>
+    public class Recursive<A>
+    {
+        private enum State
+        {
+            Recurse,
+            Return
+        }
+
+        private readonly State recursiveState;
+        private readonly Func<Recursive<A>> recursiveFunction;
+        private readonly A val;
+
+        internal Recursive(Some<A> val)
```

Comment:
Please remove `Some<A>`, it isn't used throughout language-ext due to its limitations.  Please just check the `val` for `null` and `throw`